### PR TITLE
fixed cursor position on tidy

### LINKED ISSUE #1223: Keep cursor position when tidying code
#### Nature of issue?

- Existing feature enhancement

#### Feature enhancement details:
The tidying code feature resets the cursor back to the beginning of the file that was tidied. For long files, this can be irksome, because you then have to scroll back to your position. It would be helpful to keep the cursor position from before the tidying.

### PR DESCRIPTION
**Pull request by ashu8912**
I have verified that this pull request:

* [X] has no linting errors (`npm run lint`)
* [X] is from a uniquely-named feature branch and has been rebased on top of the latest master. (If I was asked to make more changes, I have made sure to rebase onto master then too)
* [X] is descriptively named and links to an issue number, i.e. `Fixes #1223 `
Cursor Position Before Tidying
Screenshot 2020-01-15 at 4 34 12 PM
Cursor Position After Tidying
Screenshot 2020-01-15 at 4 39 36 PM

### CODE REVIEW

**Comment by jpgualdarrama** [comment]
@ashu8912 I noticed that in the "after" picture, there is indentation on line 2, but not on line 6. Is there a reason for that? I'm more of a user and not a developer, but I would expect that the indentation would be on line 6 as well.

**Comment by ashu8912** [comment]
@jpgualdarrama yes you are right it is very strange that i am just setting the cursor position without doing any harm to the beautification but somehow the indentation is getting removed will have to look into this more sending a commit soon that will fix this.

**Comment by ashu8912** [comment]
Before 
Screenshot 2020-01-25 at 6 27 32 PM

After tidy code
Screenshot 2020-01-25 at 6 29 00 PM

File: client/modules/IDE/components/Editor.jsx [hunk 1]
```diff
@@ -269,17 +269,21 @@ class Editor extends React.Component {
       indent_size: INDENTATION_AMOUNT,
       indent_with_tabs: IS_TAB_INDENT
     };
-
+    console.log('inside');
```
Comment (MichaelPaulukonis) [inline client/modules/IDE/components/Editor.jsx]: Could you remove the `console.log` statements?

File: client/modules/IDE/components/Editor.jsx [hunk 2]
```diff
@@ -269,17 +269,21 @@ class Editor extends React.Component {
...
     if (mode === 'javascript') {
       this._cm.doc.setValue(beautifyJS(this._cm.doc.getValue(), beautifyOptions));
+      console.log('beautifying');
     } else if (mode === 'css') {
       this._cm.doc.setValue(beautifyCSS(this._cm.doc.getValue(), beautifyOptions));
     } else if (mode === 'htmlmixed') {
       this._cm.doc.setValue(beautifyHTML(this._cm.doc.getValue(), beautifyOptions));
     }
-    this._cm.doc.setCursor({ line: currentPosition.line, ch: currentPosition.ch });
+    setImmediate(() => {
```
Comment (MichaelPaulukonis) [inline client/modules/IDE/components/Editor.jsx]: Why use `setImmediate` here?

MDN notes that [it is non-standard](https://developer.mozilla.org/en-US/docs/Web/API/Window/setImmediate) and should not be used.

Comment (ashu8912) [reply client/modules/IDE/components/Editor.jsx]: The problem is to set the cursor in the next tick of the event loop and setImmediate executes immediately on the nextTick But as you mentioned setImmediate is a non-standard now replacing setImmediate with setTimeOut will be a good idea

File: client/modules/IDE/components/Editor.jsx (hunk 1 above)
Comment (ashu8912) [reply client/modules/IDE/components/Editor.jsx]: thanks for mentioning i have made a commit that removes it

File: client/modules/IDE/components/Editor.jsx [hunk 3]
```diff
@@ -269,15 +269,20 @@ class Editor extends React.Component {
       indent_size: INDENTATION_AMOUNT,
       indent_with_tabs: IS_TAB_INDENT
     };
-
     const mode = this._cm.getOption('mode');
+    const currentPosition = this._cm.doc.getCursor();
     if (mode === 'javascript') {
       this._cm.doc.setValue(beautifyJS(this._cm.doc.getValue(), beautifyOptions));
+      console.log('beautifying');
```
Comment (catarak) [inline client/modules/IDE/components/Editor.jsx]: could you remove this last `console.log`? otherwise this is working great for me!

Comment (ashu8912) [reply client/modules/IDE/components/Editor.jsx]: hey @catarak this is fixed and can be merged now

**Comment by catarak** [comment]
thanks, looks great!

**Pull request merged**